DDP Communication hook: Fix the way we pass future result to buckets.

sinannasir opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#43307 DDP Communication hook: Fix the way we pass future result to buckets.**

I identified a bug with DDP communication hook while I was trying accuracy benchmarks: I was getting `loss=nan`. 

Looks like when we re-`initialize_bucketviews` with the value of `future_work`, as `Reducer::mark_variable_ready_dense` does `bucket_view.copy_(grad)` it wasn't copying the `grads` back to the contents since `bucket_view` wouldn't have any relationship with `contents` after re-intitializing it with something else. As we have multiple iterations, this was causing problems. 
I solved this by adding two states for `bucket_view`:
```
    // bucket_views_in[i].copy_(grad) and
    // grad.copy_(bucket_views_out[i])
    // provide convenient ways to move grad data in/out of contents.
    std::vector<at::Tensor> bucket_views_in;
    std::vector<at::Tensor> bucket_views_out;
```

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) 

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

sinannasir committed:
DDP Communication hook: Fix the way we pass future result to buckets.

I identified a bug with DDP communication hook while I was trying accuracy benchmarks: I was getting `loss=nan`.

Looks like passing the result of future as re-`initialize_bucketviews` was causing the problem.

One easy fix is to simply do double `copy_` by `bucket.replicas[i].contents.copy_(future_result[i]);`.

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) This simple fix just fixes the issue by doing double `copy_`. From my recent observations, performance regression by using double `copy_` is negligible. If a better solution is not obvious, we should probably land this quickly and open a separate issue to think more on the problem and a better solution.

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

[ghstack-poisoned]

rohan-varma commented on test/distributed/test_c10d.py:
```diff
@@ -2576,20 +2576,26 @@ def check_no_grads():
+        """
         int_devices = gpus_for_rank(self.world_size)[self.rank][:1]
-        devices = list([torch.device('cuda:' + str(i)) for i in int_devices])
+        devices = list([torch.device("cuda:" + str(i)) for i in int_devices])
```
nit: since you're already changing this line, might as well remove the call to `list()` which is not needed

sinannasir replied:
Oh, I missed that. I was just running linter for the parts that I was changing. I also made relevant changes across the whole test file.

pritamdamania87 reviewed (changes requested):
As discussed offline, let's figure out why the Nan is occurring and whether its possible to avoid this additional copy.

sinannasir committed:
Update on "DDP Communication hook: Fix the way we pass future result to buckets."


I identified a bug with DDP communication hook #40848 while I was trying accuracy benchmarks: I was getting `loss=nan`.

Looks like passing the result of future as re-`initialize_bucketviews` was causing the problem.

One easy fix is to simply do double `copy_` by `bucket.replicas[i].contents.copy_(future_result[i]);`.

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) This simple fix just fixes the issue by doing double `copy_`. From my recent observations, performance regression by using double `copy_` is negligible. If a better solution is not obvious, we should probably land this quickly and open a separate issue to think more on the problem and a better solution.

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

[ghstack-poisoned]

sinannasir committed:
Update on "DDP Communication hook: Fix the way we pass future result to buckets."


I identified a bug with DDP communication hook while I was trying accuracy benchmarks: I was getting `loss=nan`. 

Looks like when we re-`initialize_bucketviews` with the value of `future_work`, as `Reducer::finalize_bucket_dense` does `grad.copy_(bucket_view)` it wasn't copying the `grads` back to the contents since `bucket_view` wouldn't have any relationship with `contents` after re-intitializing it with something else. As we have multiple iterations, this was causing problems. 
I solved this by adding two states for `bucket_view`:
```[c++]
    // bucket_views_in[i].copy_(grad) and
    // grad.copy_(bucket_views_out[i])
    // provide convenient ways to move grad data in/out of contents.
    std::vector<at::Tensor> bucket_views_in;
    std::vector<at::Tensor> bucket_views_out;
```

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) This simple fix just fixes the issue by doing double `copy_`. From my recent observations, performance regression by using double `copy_` is negligible. If a better solution is not obvious, we should probably land this quickly and open a separate issue to think more on the problem and a better solution.

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

[ghstack-poisoned]

sinannasir committed:
Update on "DDP Communication hook: Fix the way we pass future result to buckets."


I identified a bug with DDP communication hook while I was trying accuracy benchmarks: I was getting `loss=nan`. 

Looks like when we re-`initialize_bucketviews` with the value of `future_work`, as `Reducer::finalize_bucket_dense` does `grad.copy_(bucket_view)` it wasn't copying the `grads` back to the contents since `bucket_view` wouldn't have any relationship with `contents` after re-intitializing it with something else. As we have multiple iterations, this was causing problems. 
I solved this by adding two states for `bucket_view`:
```[c++]
    // bucket_views_in[i].copy_(grad) and
    // grad.copy_(bucket_views_out[i])
    // provide convenient ways to move grad data in/out of contents.
    std::vector<at::Tensor> bucket_views_in;
    std::vector<at::Tensor> bucket_views_out;
```

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) This simple fix just fixes the issue by doing double `copy_`. From my recent observations, performance regression by using double `copy_` is negligible. If a better solution is not obvious, we should probably land this quickly and open a separate issue to think more on the problem and a better solution.

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

[ghstack-poisoned]

sinannasir commented:
> As discussed offline, let's figure out why the Nan is occurring and whether its possible to avoid this additional copy.

Looks like when we re-`initialize_bucketviews` with the value of `future_work`, as `Reducer::mark_variable_ready_dense` does `bucket_view.copy_(grad)` it wasn't copying the `grads` back to the contents since `bucket_view` wouldn't have any relationship with `contents` after re-intitializing it with something else. As we have multiple iterations, this was causing problems. 
I solved this by adding two states for `bucket_view`:
```
    // bucket_views_in[i].copy_(grad) and
    // grad.copy_(bucket_views_out[i])
    // provide convenient ways to move grad data in/out of contents.
    std::vector<at::Tensor> bucket_views_in;
    std::vector<at::Tensor> bucket_views_out;
```

I see that there is a recent PR on bucket_views #41954. I've solved a lot of merge conflicts with that PR. Hope my changes won't cause any trouble.

sinannasir committed:
Update on "DDP Communication hook: Fix the way we pass future result to buckets."


I identified a bug with DDP communication hook while I was trying accuracy benchmarks: I was getting `loss=nan`. 

Looks like when we re-`initialize_bucketviews` with the value of `future_work`, as `Reducer::mark_variable_ready_dense` does `bucket_view.copy_(grad)` it wasn't copying the `grads` back to the contents since `bucket_view` wouldn't have any relationship with `contents` after re-intitializing it with something else. As we have multiple iterations, this was causing problems. 
I solved this by adding two states for `bucket_view`:
```
    // bucket_views_in[i].copy_(grad) and
    // grad.copy_(bucket_views_out[i])
    // provide convenient ways to move grad data in/out of contents.
    std::vector<at::Tensor> bucket_views_in;
    std::vector<at::Tensor> bucket_views_out;
```

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) This simple fix just fixes the issue by doing double `copy_`. From my recent observations, performance regression by using double `copy_` is negligible. If a better solution is not obvious, we should probably land this quickly and open a separate issue to think more on the problem and a better solution.

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -761,9 +762,10 @@ void Reducer::initialize_buckets(
     Reducer::BucketReplica& replica,
-    at::Tensor& contents) {
+    at::Tensor& contents,
+    bool from_initialize_buckets) {
```
nit: rename this to `populate_bucket_views_in`

pritamdamania87 reviewed (approved):
Thanks for fixing this, looks good to me.

sinannasir committed:
Update on "DDP Communication hook: Fix the way we pass future result to buckets."


I identified a bug with DDP communication hook while I was trying accuracy benchmarks: I was getting `loss=nan`. 

Looks like when we re-`initialize_bucketviews` with the value of `future_work`, as `Reducer::mark_variable_ready_dense` does `bucket_view.copy_(grad)` it wasn't copying the `grads` back to the contents since `bucket_view` wouldn't have any relationship with `contents` after re-intitializing it with something else. As we have multiple iterations, this was causing problems. 
I solved this by adding two states for `bucket_view`:
```
    // bucket_views_in[i].copy_(grad) and
    // grad.copy_(bucket_views_out[i])
    // provide convenient ways to move grad data in/out of contents.
    std::vector<at::Tensor> bucket_views_in;
    std::vector<at::Tensor> bucket_views_out;
```

I included two additional unit tests where we run multiple iterations for better test coverage:
1) `test_accumulate_gradients_no_sync_allreduce_hook`
2) `test_accumulate_gradients_no_sync_allreduce_with_then_hook`.

Those tests were failing with the old version. (If I run it for just 2 iterations, it wouldn't fail.) This simple fix just fixes the issue by doing double `copy_`. From my recent observations, performance regression by using double `copy_` is negligible. If a better solution is not obvious, we should probably land this quickly and open a separate issue to think more on the problem and a better solution.

Differential Revision: [D23229309](https://our.internmc.facebook.com/intern/diff/D23229309/)

[ghstack-poisoned]

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -772,14 +774,24 @@ void Reducer::initialize_bucketviews(
       // the autograd engine (AccumulateGrad) will also create gradients
       // matching its layout.
-      replica.bucket_views.push_back(
+      replica.bucket_views_out.push_back(
```
Sorry that I didn't get a chance to look closely at this PR earlier, but I was wondering in the regular case (no comm. hoook) if bucket_views_out and bucket_views_in are essentially the same thing? If so, I was wondering if it would work for bucket_views_out to just point to bucket_views_in when there is no comm. hook?

sinannasir replied:
Thanks for spotting that. Please see #43734

rohan-varma commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -772,14 +774,24 @@ void Reducer::initialize_bucketviews(
+      // If calling from `initialize_buckets`, push_back to views_in too.
+      if (populate_bucket_views_in) {
+        replica.bucket_views_in.push_back(
+            contents.as_strided(v.sizes(), v.strides(), offset));
```
I'm not sure about the difference, but should this be `contents.narrow(...)` as the above?

sinannasir replied:
Thanks for spotting that. Please see #43734
